include/quda_matrix.h

```diff
@@ -115,18 +117,73 @@ namespace quda {
 	}
 
 	template<typename S>
-	  __device__ __host__ inline Matrix(const gauge_wrapper<typename RealType<T>::type, S> &s);
+	  __device__ __host__ inline Matrix(const gauge_wrapper<real, S> &s);
 
 	template<typename S>
-	  __device__ __host__ inline void operator=(const gauge_wrapper<typename RealType<T>::type, S> &s);
+	  __device__ __host__ inline void operator=(const gauge_wrapper<real, S> &s);
 
 	template<typename S>
-	  __device__ __host__ inline Matrix(const gauge_ghost_wrapper<typename RealType<T>::type, S> &s);
+	  __device__ __host__ inline Matrix(const gauge_ghost_wrapper<real, S> &s);
 
 	template<typename S>
-	  __device__ __host__ inline void operator=(const gauge_ghost_wrapper<typename RealType<T>::type, S> &s);
+	  __device__ __host__ inline void operator=(const gauge_ghost_wrapper<real, S> &s);
+
+        /**
+           @brief Compute the matrix L1 norm - this is the maximum of
+           the absolute column sums.
+           @return Compute L1 norm
+        */
+        __device__ __host__ inline real L1() {
+          real l1 = 0;
+#pragma unroll
+          for (int j=0; j<N; j++) {
+            real col_sum = 0;
+#pragma unroll
+            for (int i=0; i<N; i++) {
+              col_sum += abs(data[i*N + j]);
+            }
+            l1 = col_sum > l1 ? col_sum : l1;
+          }
+          return l1;
+        }
+
+        /**
+           @brief Compute the matrix L2 norm.  We actually compute the
+           Frobenius norm which is an upper bound on the L2 norm.
+           @return Computed L2 norm
+        */
+        __device__ __host__ inline real L2() {
+          real l2 = 0;
+#pragma unroll
+          for (int j=0; j<N; j++) {
+#pragma unroll
+            for (int i=0; i<N; i++) {
+              l2 += norm(data[i*N + j]);
+            }
+          }
+          return sqrt(l2);
+        }
+
+        /**
+           @brief Compute the matrix L1 norm - this is the maximum of
```

Comment:
Trivial change---this is the Linf norm, not the L1 norm! Came along with the copy+paste ride, I assume.